x-pack/plugin/kql/src/main/java/org/elasticsearch/xpack/kql/query/KqlQueryBuilder.java

```diff
@@ -0,0 +1,201 @@
+        PARSER.declareBoolean(optionalConstructorArg(), CASE_INSENSITIVE_FIELD);
+        PARSER.declareString(optionalConstructorArg(), TIME_ZONE_FIELD);
+        PARSER.declareString(optionalConstructorArg(), DEFAULT_FIELD_FIELD);
+        declareStandardFields(PARSER);
+    }
+
+    private final String query;
+    private boolean caseInsensitive = true;
+    private ZoneId timeZone;
+    private String defaultField;
+
+    public KqlQueryBuilder(String query) {
+        this.query = Objects.requireNonNull(query, "query can not be null");
+    }
+
+    public KqlQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        query = in.readString();
+        caseInsensitive = in.readBoolean();
+        timeZone = in.readOptionalZoneId();
+        defaultField = in.readOptionalString();
+    }
+
+    public static KqlQueryBuilder fromXContent(XContentParser parser) {
+        try {
+            return PARSER.apply(parser, null);
+        } catch (IllegalArgumentException e) {
+            throw new ParsingException(parser.getTokenLocation(), e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.KQL_QUERY_ADDED;
+    }
+
+    public String queryString() {
+        return query;
+    }
+
+    public boolean caseInsensitive() {
+        return caseInsensitive;
+    }
+
+    public KqlQueryBuilder caseInsensitive(boolean caseInsensitive) {
+        this.caseInsensitive = caseInsensitive;
+        return this;
+    }
+
+    public ZoneId timeZone() {
+        return timeZone;
+    }
+
+    public KqlQueryBuilder timeZone(String timeZone) {
+        this.timeZone = timeZone != null ? ZoneId.of(timeZone) : null;
+        return this;
+    }
+
+    public String defaultField() {
+        return defaultField;
+    }
+
+    public KqlQueryBuilder defaultField(String defaultField) {
+        this.defaultField = defaultField;
+        return this;
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        {
+            builder.field(QUERY_FIELD.getPreferredName(), query);
+            builder.field(CASE_INSENSITIVE_FIELD.getPreferredName(), caseInsensitive);
+
+            if (defaultField != null) {
+                builder.field(DEFAULT_FIELD_FIELD.getPreferredName(), defaultField);
+            }
+
+            if (timeZone != null) {
+                builder.field(TIME_ZONE_FIELD.getPreferredName(), timeZone.getId());
+            }
+
+            boostAndQueryNameToXContent(builder);
+        }
+        builder.endObject();
+    }
+
+    @Override
+    protected QueryBuilder doIndexMetadataRewrite(QueryRewriteContext context) throws IOException {
+        KqlParser parser = new KqlParser();
+        QueryBuilder rewrittenQuery = parser.parseKqlQuery(query, createKqlParserContext(context));
+
+        if (log.isTraceEnabled()) {
```

Comment:
Why do you keep checking `isTraceEnabled`? If you don't want to eagerly execute the `toString` use a message supplier `log.trace(() -> ...)`

Reply:
I removed the conditional and use suppliers when I want to evaluate the message lazily. Thanks for the suggestion.